### PR TITLE
Update subset

### PR DESCRIPTION
**Pull request by iattempt**
**Fixes issue:** #[Mention the issue number it fixes or add the details of the changes if it doesn't has a specific issue.]

**Changes:**
1. separate test file
2. substitute algorithm
3. remove coupling of container and algorithm

### CODE REVIEW

File: code/dynamic_programming/subset_sum/subset_sum.cpp [hunk 1]
```diff
@@ -1,44 +1,76 @@
...
-int reachable[MAX_SUM];
-bool dp[MAX_SUM];
+// complexity: time: O(sum * n), space: O(sum)
+template<typename _BidirectionalIter,
+         typename _ValueType = typename std::iterator_traits<_BidirectionalIter>::value_type>
+bool
+isSubsetSum(_BidirectionalIter begin, _BidirectionalIter end, _ValueType sum)
+{
+    auto sz = std::distance(begin, end);
+    bool subset[2][sum + 1];
```
Comment (arnavb) [inline code/dynamic_programming/subset_sum/subset_sum.cpp]: Any reason why you are using a 2D raw array?

File: code/dynamic_programming/subset_sum/subset_sum_test.cpp [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+#ifndef SUBSET_SUM_TEST
+#define SUBSET_SUM_TEST
+
+#include <iostream>
+#include <list>
+#include <cassert>
+#include "subset_sum.cpp"
+
+using namespace std;
```
Comment (arnavb) [inline code/dynamic_programming/subset_sum/subset_sum_test.cpp]: Don't use this.

Comment (iattempt) [reply code/dynamic_programming/subset_sum/subset_sum_test.cpp]: this just is test file.

File: code/dynamic_programming/subset_sum/subset_sum.cpp (hunk 1 above)
Comment (iattempt) [reply code/dynamic_programming/subset_sum/subset_sum.cpp]: [Refer to the description of this article.](http://www.geeksforgeeks.org/subset-sum-problem-osum-space/)

**Comment by AdiChat** [comment]
Looks great 😍 

Hope you are enjoying your journey 🚆  with **OpenGenus**

**Pull request merged**